feat: add spaces to color knob

francescovallone opened the pull request:
This PR adds color spaces such as HEX, RGBA, HSL, and a Color Picker with 16 colors to the color knob.

All the color spaces have their own algorithm to convert from and to HEX.

### List of issues which are fixed by the PR
#368 

### Screenshots
![photo_2023-10-26_17-14-34](https://github.com/widgetbook/widgetbook/assets/15020233/680d0e64-2114-41be-8053-0cd98d159752)
![photo_2023-10-26_17-14-32](https://github.com/widgetbook/widgetbook/assets/15020233/e2f6e679-7138-47c5-ab6f-93cb55a58a0c)
![image_2023-10-26_17-13-26](https://github.com/widgetbook/widgetbook/assets/15020233/75254967-7191-4de4-8857-523a5249c1ea)


### Checklist

- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making].
- [x] All existing and new tests are passing.

YoussefRaafatNasry reviewed (changes requested):
Great job, @francescovallone! This will be a great addition to the Color Field!

I cannot review this due to how the files are structured, please consider the following points to make the review process easier:
- Remove color picker functionality, we prefer a simple approach for now.
-  Move all widgets inside `fields/color_field.dart` all to their own file inside `fields/color_field/widget_name_here.dart`. The file `color_field.dart` should be inside the new folder as well.
- Check out all `pubspec.lock` files from the PR, you should not add these type of changes to the PR as they are heavily dependent on your local version of dart/flutter.

francescovallone commented:
@YoussefRaafatNasry oh sorry about that! I had guessed that the file structure wasn't good but since I haven't seen any other knob behaving like this one I didn't know how to manage this situation.

I will fix all the issues this afternoon 😄

YoussefRaafatNasry commented:
@francescovallone No need for an apology and tyt! The color field is a bit odd with all the custom widgets it needs, so makes sense to structure it differently.

YoussefRaafatNasry reviewed (changes requested):
Great work @francescovallone! 💙 
I think we could work a bit on improving the implementation because some classes like (`ColorsConverter`) have pretty intense logic that can be minimized.

8 months ago, before I joined Widgetbook, I was working on the same issue back then.
I didn't have the time since then to get it here. But I would advice you to have a look at this [implementation](https://github.com/YoussefRaafatNasry/color_picker) and take some naming/implementation inspiration from it.

francescovallone commented:
@YoussefRaafatNasry thank you for the review! I really appreciate it.
I will fix all the issues as soon as possible 😄!

francescovallone commented:
Hello @YoussefRaafatNasry! I addressed all the issues that you pointed out. Also, a huge thank you for making me discover the HSLColor class in Flutter it simplified a lot of logic inside this knob and also made me decide to remove altogether the `ColorConverter` class. Let me know if there are other issues 😄

JAkrish29 commented:
If the issue is still open do assign me

francescovallone commented:
Hello! Sorry for the spam but I would like to know if there are any issue or any update regarding this pr 😄.

YoussefRaafatNasry commented:
Hello @francescovallone 👋 
There's nothing blocking the PR except the review. I'm currently working on side-tasks for Widgetbook 4 that are time-consuming. Since this PR has relatively a lot of files changed/added, then it needs to be carefully reviewed.

I will get this merged before the end of this week.
So apology for the delay, and thanks again for you contribution and patience 💙

YoussefRaafatNasry commented on packages/widgetbook/lib/src/fields/color_field/color_field.dart:
```diff
@@ -0,0 +1,223 @@
+
+  final ColorSpace colorSpace;
+  final String name;
+  final Color? value;
+  final Color? initialValue;
```
We do not need both these fields, only one of them would be sufficient. Do you agree?

YoussefRaafatNasry commented on packages/widgetbook/lib/src/fields/color_field/color_space.dart:
```diff
@@ -0,0 +1,5 @@
+enum ColorSpace {
+  rgba,
```
Let's rename it to `rgb`, because the alpha component will be shared with all color space. For example, the usual `hex` color space does not have the alpha value by default.

YoussefRaafatNasry commented on packages/widgetbook/lib/src/fields/color_field/color_field.dart:
```diff
@@ -0,0 +1,223 @@
+  }
+
+  @override
+  Widget build(BuildContext context) {
```
The `ColorPicker` widget will be responsible for the alpha value, and it will no longer be the responsibly of the sub-widgets.

Here's how the design should look like:

![image](https://github.com/widgetbook/widgetbook/assets/41103290/8aa20e61-5c11-41cb-a91b-cbebf8783346)

The second row will change depending on the ColorSpace picked.

YoussefRaafatNasry commented on packages/widgetbook/lib/src/fields/color_field/color_text_field.dart:
```diff
@@ -0,0 +1,41 @@
+import 'package:flutter/material.dart';
+import 'package:flutter/services.dart';
+
+class ColorTextField extends StatelessWidget {
```
This can be refactored into `NumberTextField` as it's always used for numbers.
Then it can also be used for the alpha percentage.

YoussefRaafatNasry commented on packages/widgetbook/lib/src/fields/color_field/hsl_color_text_fields.dart:
```diff
@@ -0,0 +1,120 @@
+
+import 'color_text_field.dart';
+
+class HslColorTextFields extends StatelessWidget {
```
The comments here will apply to all other similar widgets as well (i.e. RGB and HEX widgets).

1. Should be renamed to `HslColorPicker`.
1. Should be converted to `StatefulWidget`, since `colorValue` is being manipulated.
1. `onChanged` should send out `Color` and not `List<String>`.
1. No need to assign the `key` of the `TextField`s.


You can have a look at [this widget](https://github.com/YoussefRaafatNasry/color_picker/blob/main/lib/src/pickers/hsl_color_picker.dart) for more info.

YoussefRaafatNasry reviewed (changes requested):
I tried to be as specific as possible with this review, as my past reviews were more generic due to the lack of time to review this.

The changes proposed will make the code much simpler to read & test. Please let me know if there's any comment that's unclear.

After you are done with the changes, I might commit some nitpicks to the PR and get it merged.

Thanks again for your patience, @francescovallone 💙

francescovallone commented:
Hi @YoussefRaafatNasry, thank you for your patience and time! 

I really appreciated your comments on my code and your review made me understand many of the errors I tried to fix in my last commit. 

Also, thank you for providing me with your implementation of `color_picker`, I learned a lot from it!

I hope that my latest changes resolve all the issues you pointed out in my code 😄!

As always, let me know if there are any problems!

YoussefRaafatNasry commented:
I would be grateful if you could update the screenshots in the PR description as well 🙏

francescovallone commented:
Hi @YoussefRaafatNasry, I'm glad my last changes were good enough and my PR was approved. But, not so happy that the tests failed because of this addition 😢 .

Anyway, I have updated the screenshots, let me know if you need help with anything 😄 .

YoussefRaafatNasry commented:
@francescovallone, don't worry about it.
I am still working on some more changes, so I will fix the tests when I am done with all changes.

YoussefRaafatNasry commented:
Hello @francescovallone! 👋 
Please reach us out on [this email](mailto:devrel@widgetbook.io) to claim your Hacktoberfest swag!
